### PR TITLE
Add support for installing ESM module nodes

### PR DESCRIPTION
**Pull request by hlovdal**
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

Currently node-red fails when attempting to install a node written as an ESM module, e.g.

```text
19 Sep 18:59:53 - [info] Server now running at http://127.0.0.1:1880/
19 Sep 18:59:53 - [info] Starting flows
19 Sep 18:59:53 - [info] Started flows
19 Sep 19:01:05 - [info] Installing module: @hlovdal/node-red-lowercase-in-typescript, version: 1.0.0
19 Sep 19:01:12 - [info] Installed module: @hlovdal/node-red-lowercase-in-typescript
19 Sep 19:01:12 - [info] Added node types:
19 Sep 19:01:12 - [info]  - @hlovdal/node-red-lowercase-in-typescript:lower-case : Error [ERR_REQUIRE_ESM]: require() of ES Module .../.node-red/node_modules/@hlovdal/node-red-lowercase-in-typescript/src/lower-case.js from .../node-red/packages/node_modules/@node-red/registry/lib/loader.js not supported.
Instead change the require of lower-case.js in .../node-red/packages/node_modules/@node-red/registry/lib/loader.js to a dynamic import() which is available in all CommonJS modules.
```

This pull request fixes that by using [dynamic `import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) instead of `require`.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the [forum](https://discourse.nodered.org/t/adding-support-for-installing-esm-module-nodes/81438)/slack team.
- [x] I have run `npm run test` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

**Comment by hlovdal** [comment]
I am not sure why the EasyCLA status is not updated, it is fine in [this pull request](https://github.com/node-red/node-red/pull/4356).

**Comment by hlovdal** [comment]
The errors described below were fixed independently in [pull request 4367](https://github.com/node-red/node-red/pull/4367).

------------

### TL;DR

The test failure that occurred on my CI build is not related to my changes, it is inherited from the existing code and applies to both `dev` and `master`.

### Test results & conclusion

I have now run testing overnight and the results are:

```bash
$ wc -l success-dev failure-dev
 1312 success-dev
   86 failure-dev
 1398 total
$ wc -l success-installing_esm_nodes failure-installing_esm_nodes
 1297 success-installing_esm_nodes
  103 failure-installing_esm_nodes
 1400 total
$ wc -l success-master failure-master
  416 success-master
   37 failure-master
  453 total
$
```

Or in percentages, the test fails in 6.2%, 7.4% and 8.2% of the times on the various branches, and the total failure rate for all is 7.0%.

The value that fails is almost 200 which it is compared agains, over half is 199 (details in the last step).

So conclusion, the timing granularity/accuracy of this failing test is not good enough and it will fail from time to time (around 7% of the time).


### Details

Step one was to limit testing to just the file that triggered the failure by modifying `Gruntfile.js` to have

```text
            all: { src: ["test/nodes/core/function/10-function_spec.js"]},
            core: { src: ["test/nodes/core/function/10-function_spec.js"]},
            nodes: { src: ["test/nodes/core/function/10-function_spec.js"]}
```

Step two: run the tests through a script that records the success or failure from each test:

```bash
#!/bin/sh

touch success-$1 failure-$1
while true
do
        if npm run test
        then
                date >> success-$1
        else
                date >> failure-$1
        fi
        echo ====================================================
        wc -l success-$1 failure-$1
        echo ====================================================
done
```

Step three was to test on all three branches:

```bash
timeout 8h ./run.sh installing_esm_nodes; git checkout dev; timeout 8h ./run.sh dev; git checkout master; ./run.sh master
```

Step four, analyse failure values.

To be able to see what the failure values were I ran the testing inside a [`script`](https://www.man7.org/linux/man-pages/man1/script.1.html) session, capturing text output to a `typescript` file.

```bash
$ awk '/Message received before init comple/{ A[$9]++ } END {total=0; for (i in A) {total += A[i]} for (i in A) {printf "%d %3d %5.1f%%\n", i, A[i], 100*A[i]/total; }}' typescript 
197   1   0.4%
198  88  38.9%
199 121  53.5%
200  16   7.1%
$
```

(the typescript file includes a few more tests than those captured in the `failure-*` files for those of you double checking the numbers)

**Comment by hlovdal** [comment]
Same results with node.js v18.17.1:

```text
  1260 node18-success-installing_esm_nodes
    73 node18-failure-installing_esm_nodes
  1244 node18-success-dev
    90 node18-failure-dev
  1241 node18-success-master
    91 node18-failure-master
  3999 total
```

File: packages/node_modules/@node-red/registry/lib/loader.js [hunk 1]
```diff
@@ -344,30 +344,19 @@ function loadNodeSet(node) {
...
+                // CJS will always have default, ESM might have default. (details in note below)
+                let r = module_namespace_object.default ? module_namespace_object.default : module_namespace_object;
+                // Babel related workaround. (references in note below)
+                r = r.__esModule ? r.default : r;
+                return createLoadPromise(node, r);
+            })
+            .catch(function(err) {
+                node.err = err;
+                return Promise.resolve(node);
+            });
```
Comment (knolleary) [inline packages/node_modules/@node-red/registry/lib/loader.js:359]:
```suggestion
        return importPromise
            .then(function(moduleNamespaceObject) {
                // CJS will always have default, ESM might have default. (details in note below)
                let r = moduleNamespaceObject.default ? moduleNamespaceObject.default : moduleNamespaceObject;
                // Babel related workaround. (references in note below)
                r = r.__esModule ? r.default : r;
                return createLoadPromise(node, r);
            })
            .catch(function(err) {
                node.err = err;
                return Promise.resolve(node);
            });
```

**Review by fbuys — COMMENTED** [review]
I am excited for this to be released, I left a comment if it helps.

File: packages/node_modules/@node-red/registry/lib/loader.js [hunk 2]
```diff
@@ -344,30 +344,19 @@ function loadNodeSet(node) {
...
-                });
-            }
-        }
-        if (loadPromise == null) {
-            node.enabled = true;
-            node.loaded = true;
-            loadPromise = Promise.resolve(node);
-        }
-        return loadPromise;
+        let importPromise = import(node.file);
```
[the remark this answers is not in the document]
Comment (hlovdal) [reply packages/node_modules/@node-red/registry/lib/loader.js:353]: The `loadNodeSet` function is supposed to return a promise, which it currently does by chaining an import promise with a load promise. I do not see how an async IIFE is useful here.

**Comment by hlovdal** [comment]
The original branch before the rebase is saved in my repo as `installing_esm_nodes.2023`.

**Review by Julian-Sz — APPROVED** [review]
Looking forward to this👍

**Comment by hlovdal** [comment]
Argh, the stupid update branch button has a default action to merge dev if you miss the drop down indicator on the side by a few pixels. Of course I intended to rebase. I'll fix it manually with a force push.

**Comment by hlovdal** [comment]
@knolleary 

Can you please merge this pull request? It is 2026 and ESM is undeniably the future of javascript, and it has already been for a very, very, very long time. By not supporting nodes in ESM format node-red is preventing the world from progressing.

While ESM support [slowly is adapted for `require`](https://socket.dev/blog/node-js-25-4-0-ships-with-stable-require-esm) and this pull request perhaps is not strictly needed for V5, [V4 supports node.js v18](https://nodered.org/docs/faq/node-versions) which most certainly will not get support for that.

And even with the new esm support for require, merging this into V5 later will reduce version differences, and I would expect future undiscovered issues related to require(esm) to be much more probable than standard dynamic import which has been stable for years. But if you want this to stop in V4 and just rely on the behaviour of the runtime, that is a viable option. But for V4 this is an absolute must have functionality.

The branch is rebased to the latest V4 version and is now fully up to date. For testing you can use my [lowercase ESM node](https://github.com/hlovdal/hlovdal-node-red-lowercase-in-typescript/releases/tag/wip-esm.2026-01-27_001). Or compile yourself from the `wip/esm` branch. The lowercase repo contains a full Cypress end to end test that installs the lowercase node and tests that it works. Update the [`run-node-red.sh`](https://github.com/hlovdal/hlovdal-node-red-lowercase-in-typescript/blob/wip/esm/bin/run-node-red.sh) to point to your node-red repo (with the `installing_esm_nodes` branch checked out and build).

**Comment by Julian-Sz** [comment]
I am also very much looking forward to this, hopefully it will get merged soon... 2.5 years and not much happened so far... :(

**Comment by TotallyInformation** [comment]
There is an issue with the code I'm afraid. It will not work on Windows because `import`s require a URL, not a file path. The correct way is:

```js
const { pathToFileURL } = require('node:url');

// ....

// let importPromise = import(node.file);
const modulePath = pathToFileURL(node.file).href;
let importPromise = import(modulePath);
```

**Comment by hlovdal** [comment]
@TotallyInformation Thank you for reporting. I will look into that (although possibly not immediately the first few days).

**Comment by hlovdal** [comment]
The windows issue has been fixed for a while. On my [node-red node in typescript repo](https://github.com/hlovdal/hlovdal-node-red-lowercase-in-typescript) I have set up automatic tests for full end-2-end tests with [Cypress](https://www.cypress.io/), [installing and testing an esm node on a node-red instance build from the `installing_esm_nodes` branch](https://github.com/hlovdal/hlovdal-node-red-lowercase-in-typescript/blob/wip/esm/.github/workflows/test-esm.yml).

For now the windows runner times out because `concurrently` fails to terminate the node-red instance and I have not spent a lot of time trying to fix it. But the e2e test itself [succeeds](https://github.com/hlovdal/hlovdal-node-red-lowercase-in-typescript/actions/runs/23324736528/job/67843456793):

```text
[CYPRESS]   ┌───────────────────────────────────────────────────────────────────────────────────────────
[CYPRESS]   │ ✔  upload_and_run.cy.ts                     01:08        1        1        -        -        - │
[CYPRESS]   └───────────────────────────────────────────────────────────────────────────────────────────
[CYPRESS]     ✔  All specs passed!                        01:08        1        1        -        -        -  
[CYPRESS] 
[CYPRESS] wait-on http://localhost:1880 && npm run cy:run exited with code 0
--> Sending SIGTERM to other processes..
Error: The operation was canceled.
```

@TotallyInformation Do you want to be the author of [the commit that fixes the import](https://github.com/hlovdal/node-red/commit/a8816594b9e7d738240658f92662c6ebbd3eba4b)? It is basically identical with [your comment](https://github.com/node-red/node-red/pull/4355#issuecomment-3811382767).

**Comment by TotallyInformation** [comment]
> @TotallyInformation Do you want to be the author of [the commit that fixes the import](https://github.com/hlovdal/node-red/commit/a8816594b9e7d738240658f92662c6ebbd3eba4b)? It is basically identical with [your comment](https://github.com/node-red/node-red/pull/4355#issuecomment-3811382767).

No, that's fine. Thanks for the offer though.

**Comment by hlovdal** [comment]
All commits pass after the latest rebase.

```shell
git test add -t test "npm ci && npm run test"
git test run -t test dev..installing_esm_nodes
```

**Comment by knolleary** [comment]
@hlovdal realise I missed this for today's beta. My apologies - it's now on my radar to make sure we get it done before final 5.0 release.

**Comment by knolleary** [comment]
@hlovdal thank you for your patience. This will be in NR5.
